### PR TITLE
Add CategoricalFocalCrossentropy to Losses API

### PR DESCRIPTION
**Pull request by Frightera**
Implements the `CategoricalFocalCrossentropy()` loss based on the paper [Focal Loss for Dense Object Detection](https://arxiv.org/pdf/1708.02002.pdf) (Lin et al., 2018).

Feature request was made in #17583.

### CODE REVIEW

**Comment by Frightera** [comment]
@haifeng-jin Thanks for the feedback! I run the format.sh on the repo but I guess it does not indentation into account.

File: keras/losses.py [hunk 1]
```diff
@@ -2025,6 +2150,142 @@ def _ragged_tensor_categorical_crossentropy(
...
+      from_logits: Whether `y_pred` is expected to be a logits tensor. By
+        default, we assume that `y_pred` encodes a probability distribution.
+      label_smoothing: Float in [0, 1]. If > `0` then smooth the labels. For
+        example, if `0.1`, use `0.1 / num_classes` for non-target labels
+        and `0.9 + 0.1 / num_classes` for target labels.
+      axis: Defaults to -1. The dimension along which the entropy is
+        computed.
+
+    Returns:
+      Categorical focal crossentropy loss value.
```
Comment (haifeng-jin) [inline keras/losses.py:2276]: Please also change this part to 4 indents.

File: keras/losses.py [hunk 2]
```diff
@@ -922,6 +922,131 @@ def __init__(
...
+        loss. Default value is `AUTO`. `AUTO` indicates that the reduction
+        option will be determined by the usage context. For almost all cases
+        this defaults to `SUM_OVER_BATCH_SIZE`. When used under a
+        `tf.distribute.Strategy`, except via `Model.compile()` and
+        `Model.fit()`, using `AUTO` or `SUM_OVER_BATCH_SIZE`
+        will raise an error. Please see this custom training [tutorial](
+        https://www.tensorflow.org/tutorials/distribute/custom_training)
+        for more details.
+      name: Optional name for the instance.
+        Defaults to 'categorical_focal_crossentropy'.
```
[the remark this answers is not in the document]
Comment (haifeng-jin) [reply keras/losses.py:1013]: Sorry about the confusion. We are migrating all docstrings to 4 indents. We just didn't have enough resource to update the losses docstrings yet.

[the remark this answers is not in the document]
Comment (Frightera) [reply keras/losses.py]: Thanks, changed all of them to 4 indents.

**Comment by Frightera** [comment]
@qlzh727 Can you take a look at this PR aswell? Thanks.

**Review by haifeng-jin — APPROVED** [review]

File: keras/backend.py [hunk 3]
```diff
@@ -5574,6 +5574,88 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+    hard examples. By default, the focal tensor is computed as follows:
+
+    It has pt defined as:
+    pt = p, if y = 1 else 1 - p
+
+    The authors use alpha-balanced variant of focal loss in the paper:
+    FL(pt) = −α_t * (1 − pt)^gamma * log(pt)
+
+    Extending this to multi-class case is straightforward:
+    FL(pt) = α_t * (1 − pt)^gamma * CE, where minus comes from
```
Comment (qlzh727) [inline keras/backend.py:5601]: Please properly convert the notation here, eg it is unclear to me what is `pt` `p`, `y` and `FL`, especially `α_t`

File: keras/backend.py [hunk 4]
```diff
@@ -5574,6 +5574,88 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+    negative log-likelihood and included in CE.
```
Comment (qlzh727) [inline keras/backend.py:5602]: What is CE here?

File: keras/backend.py [hunk 5]
```diff
@@ -5574,6 +5574,88 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+
+    `modulating_factor` is (1 − pt)^gamma, where `gamma` is a focusing
+    parameter. When `gamma` = 0, there is no focal effect on the categorical
+    crossentropy. And if alpha = 1, at the same time the loss is equivalent
+    to the categorical crossentropy.
+
+    Args:
+        target: A tensor with the same shape as `output`.
```
Comment (qlzh727) [inline keras/backend.py:5610]: Target is usually the label/ground truth from the dataset.

File: keras/backend.py [hunk 6]
```diff
@@ -5574,6 +5574,88 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+        output: A tensor.
```
Comment (qlzh727) [inline keras/backend.py:5611]: Output is the model prediction.

File: keras/backend.py [hunk 7]
```diff
@@ -5574,6 +5574,88 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+    """
+    target = tf.convert_to_tensor(target)
+    output = tf.convert_to_tensor(output)
+    target.shape.assert_is_compatible_with(output.shape)
+
+    output, from_logits = _get_logits(
+        output, from_logits, "Softmax", "categorical_focal_crossentropy"
+    )
+
+    output = tf.__internal__.smart_cond.smart_cond(
```
Comment (qlzh727) [inline keras/backend.py:5634]: Just curious, `from_logits` should be a python boolean, right? why do we use smart_cond here for tensor value check?

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Thanks for the PR and sorry for the late reply. 

I think the change overall looks good, we need do some clean up for the docstring.

All my comment to the backend.py should be applied to the losses.py as well.

**Comment by qlzh727** [comment]
Adding Francois as well since this change is large and touch the API interface.

File: keras/backend.py (hunk 7 above)
Comment (Frightera) [reply keras/backend.py:5634]: That was confusing for me aswell, I sticked the one in `binary_focal_crossentropy`:

https://github.com/keras-team/keras/blob/c72e310ab5452e5a8bbcc2f18ab092f21d8846b2/keras/backend.py#L5764-L5768

If you agree, I can change this into a plain `if` statement.

File: keras/backend.py [hunk 8]
```diff
@@ -5574,6 +5574,88 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+        from_logits,
+        lambda: softmax(output),
+        lambda: output,
+    )
+
+    # scale preds so that the class probas of each sample sum to 1
```
[the remark this answers is not in the document]
Comment (Frightera) [reply keras/backend.py:5640]: Again this was taken from `categorical_crossentropy`:

https://github.com/keras-team/keras/blob/c72e310ab5452e5a8bbcc2f18ab092f21d8846b2/keras/backend.py#L5569-L5570

I can change this to: "Adjust the predictions so that the probability of each class for every sample adds up to 1."

**Comment by Frightera** [comment]
Thanks for the review @qlzh727, I'll review the clarity of the docstrings. Maybe I can make minor changes on other loss functions aswell (`binary_focal_crossentropy`, `categorical_crossentorpy`), please see my replies above.

File: keras/backend.py (hunk 7 above)
Comment (Frightera) [reply keras/backend.py]: This is changed to an if statement now

**Comment by Frightera** [comment]
@qlzh727 Re-requesting a review removed Francois, can you add him again? Also fixed the docstrings in terms of clarity. 

Can you check once more? Thanks.

**Review by fchollet — APPROVED** [review]
Looks good! Thank you.

**Comment by haifeng-jin** [comment]
@Frightera Here is what would happen next.

After @qlzh727 approval, I will start to help you migrate it. It would take some work to adapt to our internal system for anything that changes the APIs.

One more thing we need your help is to update the (tensorflow/tensorflow RELEASE.md)[https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md] to include your changes in the release notes. You may start this after this PR is completely merged.

Thanks!

**Review by qlzh727 — APPROVED** [review]

**Comment by haifeng-jin** [comment]
@Frightera I have added the release notes altogether.
So you do not need to submit a separate PR for it.
Everything will be done when this PR is merged.

**Comment by Frightera** [comment]
@haifeng-jin Thanks, appreciate it 👍🏼
